mt7621: grab macaddr straight from u-boot-env

neheb opened the pull request:
Avoids userspace handling.

not sure if this is even correct. ping @DragonBluep @rmilecki 

@musashino205 can you test if this works on that iodata device?

DragonBluep commented on target/linux/ramips/dts/mt7621_h3c_tx180x.dtsi:
```diff
@@ -86,8 +95,13 @@
 			reg = <0x0080000 0x0080000>;
+
+			macaddr_uboot_eth: ethaddr {
+				#nvmem-cell-cells = <1>;
```
Please search `package/boot/uboot-envtools/files/ramips` to find the "u-boot,env" size. If the partition size is not equal to the env size, we need to create a sub partition to wrap it. That's because the driver will check the u-env crc32 value. see: https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/tree/drivers/nvmem/u-boot-env.c?h=next-20231208#n186

i.e.
https://github.com/openwrt/openwrt/pull/13120/commits/234bf01178731a882076af298f79764feeaedffd

neheb replied:
Ah right. I have a device like this. Unfortunate.

neheb replied:
I believe this is fine now.

neheb replied:
Although I do wonder if everything should be in a subnode for consistency...

musashino205 replied:
The data length of env on the following devices is ```0x1000```.

- ELECOM WRC-2533GHBK-I
- ELECOM WRC-2533GHBK2-T
- WNPR2600G

neheb replied:
Based on my reading of https://github.com/openwrt/openwrt/commit/db93d6ca67dbdbe1b496b53c9b85d777bbce448a, the size should be the last value of ubootenv_add_uci_config

DragonBluep reviewed (commented):
We need to enable this feature first. ref: https://github.com/openwrt/openwrt/pull/13120/commits/db93d6ca67dbdbe1b496b53c9b85d777bbce448a

This will increase kernel size by ~ 1 KB.

john-tho commented:
Would need to enable the driver in kernel. Something like `CONFIG_NVMEM_U_BOOT_ENV=y` https://github.com/openwrt/openwrt/blob/main/target/linux/generic/backport-5.15/807-v6.1-0002-nvmem-add-driver-handling-U-Boot-environment-variabl.patch

---

Also would like to hear from Rafał if we should wait for potential `u-boot,env` as layout changes now [`nvmem: core: Rework layouts to become regular devices`](https://lore.kernel.org/lkml/20231129163737.698317-7-miquel.raynal@bootlin.com/) is accepted to nvmem (now waiting for push and accepted by GKH).

neheb commented:
huh. https://github.com/u-boot/u-boot/commit/8142c4554ffaa927529f24427a35f7ee2861793a is upstream now.

Interesting...

neheb commented:
questionable commit added.

rmilecki commented:
I wanted to get rid of `fw_env.config` for a long time now. We should have:
1. Kernel providing NVMEM device access
2. Ethernet drivers obtaining MAC using NVMEM cell
3. Userspace tools (`fw_printenv` / `fw_setenv`) use NVMEM device access

That would save us quite some code.

Now, I did some of the work but there is more to accomplish.

1. Allow cells other than `ethaddr` to provide MAC address
Even in this pull requests I can see some devices use `wanaddr` that won't be parsed and provided in binary MAC format to consumers.

2. Make `fw_env` full work with NVMEM devices
I pushed regression fix yesterday and I sent another patch to U-Boot community today.

3. Add support for specifying env data size
WIP. See:
* [[PATCH 1/3] dt-bindings: nvmem: u-boot,env: Add support for u-boot,env-size](https://patchwork.ozlabs.org/project/devicetree-bindings/patch/20230724082632.21133-1-ansuelsmth@gmail.com/)
* [[PATCH 1/2] dt-bindings: nvmem: add "data-size" property](https://patchwork.ozlabs.org/project/devicetree-bindings/patch/20231213154721.5037-1-zajec5@gmail.com/)
* [[PATCH 2/2] dt-bindings: nvmem: u-boot,env: describe "data-size" property](https://patchwork.ozlabs.org/project/devicetree-bindings/patch/20231213154721.5037-2-zajec5@gmail.com/)

4. Make U-Boot env an NVMEM layout
I was waiting for refactoring NVMEM drivers. It's done now but I'll wait an extra week or two for Greg to accept Srini's patches.

In other words this is great work but a bit ahead of its time. We need few more changes first to support all cases.

rmilecki commented:
I just sent RFC for specifying more MAC U-Boot environment variables (NVMEM cells):
[[PATCH RFC] dt-bindings: nvmem: u-boot,env: add any-name MAC cells compatible](https://patchwork.ozlabs.org/project/devicetree-bindings/patch/20231214153620.23998-1-zajec5@gmail.com/)

My plan is to:
1. Wait for layouts refactoring to get accepted by Greg
2. Convert U-Boot env to NVMEM layout (it'll be tricky because of backward compatibility)
3. Add support for env variables (NVMEM cells) other than `ethaddr`

neheb commented:
@rmilecki what do you think about only applying the needed dts changes and not applying the envtools config removals nor the CONFIG_ symbols for now?

rmilecki commented:
U-Boot env data should use `nvmem-layout` once it's ready (I already sent proper patches) so I'd prefer to have that handled in 1 change instead modifying the same code twice.

neheb commented:
I should probably redo this PR. label_mac is not being handled.

neheb committed:
ramips: h3x,tx180x: use nvmem for uboot-env

Removes deprecated userspace handling.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/14182
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb committed:
ramips: iptime,t5004: use nvmem for ubootenv

Userspace handling is deprecated.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/14182
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb commented:
gonna keep it simple for now so it can be merged.

neheb commented:
@redrathnure I have a question about edimax,ra21s and edimax,rg21s

original dts uses <&factory 0x4>; as mac address for gmac0 but 02_network uses mtd_get_mac_ascii u-boot-env ethaddr for gmac0 and mtd_get_mac_ascii u-boot-env wanaddr for eth1. Meaning the latter overrides the former.

redrathnure commented:
> @redrathnure I have a question about edimax,ra21s and edimax,rg21s

Sorry, I do not have the device anymore so cannot answer the question.

neheb committed:
ramips: edimax,rx21s: use nvmem for ethaddr

Userspace handling is deprecated.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/14182
Signed-off-by: Robert Marko <robimarko@gmail.com>

musashino205 commented:
> `ramips: h3x_tx180x: `

`ramips: h3c_tx180x: `?

> `ramips: wrc-2533hbk: `

`ramips: wrc-2533ghbk: `?

And, `ramips: <vendor>_<model>: ` vs. `ramips: <vendor>,<model>: ` vs. `ramips: <model>: `

neheb commented:
rebased.

@musashino205 didn't understand your comment. I think I fixed it though.

neheb commented:
Rebased and converted to nvmem-layout.

musashino205 commented:
ELECOM WRC-2533GHBK-I and WRC-2533GHBK2-T have quoted `ethaddr`, so the current NVMEM driver is not usable for them.

neheb commented:
@musashino205 removed.

neheb commented:
@musashino205 on that blog post, I see edimax. Which would imply this is an issue with these edimax devices...

neheb commented:
Rebased.

neheb committed:
ramips: elecom,wrc-2533ghbk: use nvmem for mac addresses

Userspace handling is deprecated.

Since the u-boot ethaddr variable is quoted, we cannot use it.

Use mac-base instead to specify in dts.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/14182
Signed-off-by: Robert Marko <robimarko@gmail.com>
